src/mod_muc.erl

```diff
@@ -838,61 +843,58 @@ iq_disco_items(Host, From, Lang, Rsm) ->
                                               {<<"name">>, Desc}]}};
                          _ ->
                              false
-                     end
-             end, Rooms) ++ RsmOut.
+                     end;
+                 ({{ Name, _ }, _}) ->
+                     {true,
+                     #xmlel{name = <<"item">>,
+                            attrs = [{<<"jid">>, jid:to_binary({Name, Host, <<>>})},
+                                     {<<"name">>, Name}]}
+                     }
 
+             end, Rooms) ++ RsmOut.
 
 -spec get_vh_rooms(jid:server(), jlib:rsm_in()) -> {list(), jlib:rsm_out()}.
-get_vh_rooms(Host, #rsm_in{max=M, direction=Direction, id=I, index=Index}) ->
-    AllRooms = lists:sort(get_vh_rooms(Host)),
-    Count = erlang:length(AllRooms),
-    Guard = case Direction of
-                _ when Index =/= undefined ->
-            [{'=:=', {element, 2, '$1'}, Host}];
-                aft ->
-            [{'=:=', {element, 2, '$1'}, Host},
-             {'>',   {element, 1, '$1'}, I}]; %% not exact here
-        before when I =/= <<>> ->
-            [{'=:=', {element, 2, '$1'}, Host},
-             {'<',   {element, 1, '$1'}, I}]; %% not exact here
-                _ ->
-            [{'=:=', {element, 2, '$1'}, Host}]
-            end,
-    L = lists:sort(
-          mnesia:dirty_select(muc_online_room,
-                              [{#muc_online_room{name_host = '$1', _ = '_'},
-                                Guard,
-                                ['$_']}])),
+get_vh_rooms(Host, #rsm_in{max=Max, direction=Direction, id=I, index=Index}) ->
+    Rooms = get_vh_rooms(Host) ++ get_all_vh_rooms(Host),
+    BareSortedRooms =
+        lists:usort(lists:map(
+            fun({_,Room,PidOrEmptyList}) -> {Room, PidOrEmptyList} end, Rooms)),
+    Count = erlang:length(BareSortedRooms),
+
     L2 = case {Index, Direction} of
-             {undefined, before} ->
-                 lists:reverse(lists:sublist(lists:reverse(L), 1, M));
-             {undefined, _} ->
-                 lists:sublist(L, 1, M);
-             {Index, _} when Index > Count orelse Index < 0 ->
-                 [];
-             _ ->
-                 lists:sublist(L, Index+1, M)
+        {undefined, undefined} when I == undefined ->
+            lists:sublist(BareSortedRooms, 1, Max);
+        {undefined, aft} ->
+            lists:dropwhile(
+                fun({{Id, _}, _}) -> Id =< I end,
+                BareSortedRooms);
+        {undefined,before} when I == <<>> ->
+            lists:reverse(lists:sublist(lists:reverse(BareSortedRooms), 1, Max));
+        {undefined, before} ->
+            L = lists:takewhile(
+                fun({{Id, _}, _}) -> Id < I end,
+                BareSortedRooms),
+            lists:reverse(lists:sublist(lists:reverse(L), 1, Max));
+        _ -> []
```

Comment:
I'm not so sure about this refactor. I think it breaks for RSM with index and as far as I can see it is not covered by `muc_SUITE`, so it's not verified. If you'd like to improve pagination logic, please add extra tests.

Reply:
I'll add some tests ;)